dashboard/app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -5,6 +5,46 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
 
   # GET /users/auth/:provider/callback
   def all
+    auth_params = request.env['omniauth.params']
+    if auth_params['state'] == 'connect_provider' && current_user
+      connect_provider
+    else
+      login
+    end
+  end
+
+  # Call /users/auth/google_oauth2?state=connect_provider and the callback will trigger this code path
+  def connect_provider
```

Comment:
Proposal: Let's make `connect_provider` take an `auth_hash` as an argument instead of referring to the request, which will make it easier to unit test.  We could write a separate helper that extracts the needed parameters from `request.env` into a standard format, taking provider differences into account before we get to the actual logic of creating an AuthenticationOption.